apiserver/internal/handlers/resource/resource.go

```diff
@@ -0,0 +1,417 @@
+		resourceServiceGetter: resourceServiceGetter,
+		validator:             validator,
+	}
+}
+
+// ServeHTTP implements http.Handler.
+func (h *ResourceHandler) ServeHTTP(resp http.ResponseWriter, req *http.Request) {
+	tag, err := h.authFunc(req, names.UserTagKind, names.MachineTagKind, names.ControllerAgentTagKind, names.ApplicationTagKind)
+	if err != nil {
+		if err := sendError(resp, err, h.logger); err != nil {
+			h.logger.Errorf("%v", err)
+		}
+		return
+	}
+
+	resourceService, err := h.resourceServiceGetter.Resource(req)
+	if err != nil {
+		if err := sendError(resp, err, h.logger); err != nil {
+			h.logger.Errorf("returning error to user: %v", err)
+		}
+		return
+	}
+
+	switch req.Method {
+	case "GET":
+		reader, size, err := h.download(resourceService, req)
+		if err != nil {
+			if err := sendError(resp, err, h.logger); err != nil {
+				h.logger.Errorf("%v", err)
+			}
+			return
+		}
+		defer reader.Close()
+		header := resp.Header()
+		header.Set("Content-Type", params.ContentTypeRaw)
+		header.Set("Content-Length", fmt.Sprint(size))
+		resp.WriteHeader(http.StatusOK)
+		var bytesWritten int64
+		if bytesWritten, err = io.Copy(resp, reader); err != nil {
+			if err := sendError(resp, errors.Errorf("resource download failed: %w", err), h.logger); err != nil {
+				h.logger.Errorf("returning error to user: %v", err)
+			}
+			return
+		}
+		if bytesWritten != size {
+			h.logger.Errorf("resource download size does not match expected resource size: %d != %d", bytesWritten, size)
+		}
+	case "PUT":
+		if err := h.changeAllowedFunc(req.Context()); err != nil {
+			if err := sendError(resp, err, h.logger); err != nil {
+				h.logger.Errorf("returning error to user: %v", err)
+			}
+			return
+		}
+		response, err := h.upload(resourceService, req, tagToUsername(tag))
+		if err != nil {
+			if err := sendError(resp, err, h.logger); err != nil {
+				h.logger.Errorf("returning error to user: %v", err)
+			}
+			return
+		}
+		if err := sendStatusAndJSON(resp, http.StatusOK, &response); err != nil {
+			h.logger.Errorf("sending response: %v", err)
+		}
+	default:
+		if err := sendError(resp, jujuerrors.MethodNotAllowedf("unsupported method: %q", req.Method), h.logger); err != nil {
+			h.logger.Errorf("returning error to user: %v", err)
+		}
+	}
+}
+
+func (h *ResourceHandler) download(service ResourceService, req *http.Request) (io.ReadCloser, int64, error) {
+	defer func() {
+		if req.Body != nil {
+			req.Body.Close()
+		}
+	}()
+
+	query := req.URL.Query()
+	application := query.Get(":application")
+	name := query.Get(":resource")
+
+	uuid, err := service.GetResourceUUIDByApplicationAndResourceName(req.Context(), application, name)
+	if errors.Is(err, resourceerrors.ResourceNotFound) {
+		return nil, 0, jujuerrors.NotFoundf("resource %s of application %s", name, application)
+	} else if err != nil {
+		return nil, 0, fmt.Errorf("getting resource uuid: %w", err)
+	}
+
+	res, reader, err := service.OpenResource(req.Context(), uuid)
```

Comment:
When opening a resource, does this also not return `ResourceNotFound`? I'm guessing it does, but if it doesn't it should.

Reply:
It does, but its not being used here since this case would be caught in the call above, so I've only documented the error we are using on the interface.

Reply:
On reflection have added it here, things could change between here and the previous call.

Reply:
BINGO!